tools/resource-gen/main.go

```diff
@@ -19,6 +19,126 @@ import (
 	_ "github.com/kumahq/kuma/api/mesh/v1alpha1"
 )
 
+// NativeApuTemplate for creating a Kuma resource.
+var NativeApiTemplate = template.Must(template.New("resource").Parse(`
+// Generated by tools/resource-gen
+// Run "make generate" to update this file.
+
+{{ $pkg := printf "%s_proto" .Package }}
+{{ $tk := "` + "`" + `" }}
+
+// nolint:whitespace
+package v1alpha1
+
+import (
+	"github.com/golang/protobuf/proto"
+
+	{{ $pkg }} "github.com/kumahq/kuma/api/{{ .Package }}/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/pkg/model"
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s/native/pkg/registry"
+
+)
+
+{{range .Resources}}
+{{- if not .SkipNativeGenerate }}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}} struct {
+	metav1.TypeMeta   {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ObjectMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+
+	Mesh   string                        {{ $tk }}json:"mesh,omitempty"{{ $tk }}
+{{- if eq .ResourceType "DataplaneInsight" }}
+	Status   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"status,omitempty"{{ $tk }}
+{{- else}}
+	Spec   *{{ $pkg }}.{{.ProtoType}} {{ $tk }}json:"spec,omitempty"{{ $tk }}
+{{- end}}
+}
+
+// +kubebuilder:object:root=true
+type {{.ResourceType}}List struct {
+	metav1.TypeMeta {{ $tk }}json:",inline"{{ $tk }}
+	metav1.ListMeta {{ $tk }}json:"metadata,omitempty"{{ $tk }}
+	Items           []{{.ProtoType}} {{ $tk }}json:"items"{{ $tk }}
+}
```

Comment:
I'm not sure that I'm convinced that generating the CRD is worth it. It makes it harder to add kubebuilder annotations that might be needed, and to add doc comments that will end up in the CRD.

I do agree with generating the helpers though.

Reply:
Going to leave the CRD generation in for now. It's easy enough to just remove that autogen and split up the previously autogened code back into individual files if we want in the future. Will generate the helpers in a subsequent PR.